src/imagecaption/imagecaptionediting.js

```diff
@@ -182,17 +178,39 @@ export default class ImageCaptionEditing extends Plugin {
 		const model = this.editor.model;
 		const changes = model.document.differ.getChanges();
 
+		const imagesWithoutCaption = [];
+
 		for ( const entry of changes ) {
-			if ( entry.type == 'insert' && entry.name == 'image' ) {
+			if ( entry.type == 'insert' && entry.name != '$text' ) {
 				const item = entry.position.nodeAfter;
 
-				if ( !getCaptionFromImage( item ) ) {
-					writer.appendElement( 'caption', item );
+				if ( item.is( 'image' ) && !getCaptionFromImage( item ) ) {
+					imagesWithoutCaption.push( item );
+				}
+
+				// Check elements with children for nested images.
+				if ( !item.is( 'image' ) && item.childCount ) {
+					// Use the walker to find all nested images despite of their nest level.
+					const walker = model.createRangeOn( item ).getWalker();
```

Comment:
You can use `model.createRangeOn( ... ).getItems()`. It will be easier to iterate through elements in a range.

Reply:
Also, since you already know that `item` is not an image, you can go with `createRangeIn()`. Right?

Reply:
Thanks for the tips. I didn't know about `.getItems()` and you're right about `crateRangeIn()` I suppose ;)

Fixing.